buffer/src/main/java/io/netty5/buffer/api/Buffer.java

```diff
@@ -1142,95 +1102,11 @@ default Buffer split() {
-     * <p>
-     * Unlike the {@link #forEachWritable(int, WritableComponentProcessor)} method, this API permits <em>external</em>
-     * iteration of the components, while at the same time protecting the life-cycle of the buffer.
-     * <p>
-     * The typical code pattern for using this API looks like the following:
-     * <pre>{@code
-     *      try (var iteration = buffer.forEachWritable()) {
-     *          for (var c = iteration.first(); c != null; c = c.next()) {
-     *              ByteBuffer componentBuffer = c.writableBuffer();
-     *              // ...
-     *          }
-     *      }
-     * }</pre>
-     * Note the use of the {@code var} keyword for local variables, which are required for correctly expressing the
-     * generic types used in the iteration.
-     * Following this code pattern will ensure that the components, and their parent buffer, will be correctly
-     * life-cycled.
-     * <p>
-     * <strong>Note</strong> that the {@link WritableComponent} instances exposed by the iterator could be reused for
-     * multiple calls, so the data must be extracted from the component in the context of the iteration.
-     * <p>
-     * The {@link ByteBuffer} instances obtained from the component, share lifetime with that internal component.
-     * This means they can be accessed as long as the internal memory store remain unchanged. Methods that may cause
-     * such changes are {@link #split(int)}, {@link #split()}, {@link #readSplit(int)}, {@link #writeSplit(int)},
-     * {@link #compact()}, {@link #ensureWritable(int)}, {@link #ensureWritable(int, int, boolean)},
-     * and {@link #send()}.
-     * <p>
-     * The best way to ensure this doesn't cause any trouble, is to use the buffers directly as part of the iteration.
-     * <p>
-     * <strong>Note</strong> that the arrays, memory addresses, and byte buffers exposed as components by this method,
-     * should not be used for changing the buffer contents beyond the respective array offset and length,
-     * or buffer position and limit. Doing so may cause undefined behaviour.
-     * <p>
-     * Changes to position and limit of the byte buffers exposed via the processed components, are not reflected back to
-     * this buffer instance.
-     *
-     * @return A component iterator of {@linkplain ReadableComponent readable components}.
-     * @param <T> An intersection type that presents both the {@link ReadableComponent} interface,
+     * @return A component iterator of {@linkplain BufferComponent readable components}.
+     * @param <T> An intersection type that presents both the {@link BufferComponent} interface,
      *          <em>and</em> the ability to progress the iteration via the {@link Next#next()} method.
      */
-    <T extends WritableComponent & ComponentIterator.Next> ComponentIterator<T> forEachWritable();
+    <T extends BufferComponent & ComponentIterator.Next> ComponentIterator<T> forEachComponent();
```

Comment:
I wonder if we should change this to something like:

```
R <R> forEachComponent(Function<R, T extends BufferComponent & ComponentIterator.Next> ComponentIterator<T>>)
```

This way we could make it safer to ensure that the native address "never escapes" while iteration is in progress.

Reply:
We might also want to consider adding an extra method that takes a `ToLongFunction ` to reduce the overhead of boxing/unboxing.

Reply:
This sounds like internal iteration again, where we often end up allocating objects for capturing lambdas.
And the ability to return data creates a path for the native address to escape, so it doesn't sound much safer than the try-with-resources we currently have.